runners/flink/1.15/src/main/java/org/apache/beam/runners/flink/transform/sql/SqlTransformUtils.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.flink.transform.sql;
+
+import org.apache.beam.runners.flink.translation.types.TypeInformationCoder;
+import org.apache.beam.sdk.coders.CannotProvideCoderException;
+import org.apache.beam.sdk.coders.CoderRegistry;
+import org.apache.beam.sdk.schemas.NoSuchSchemaException;
+import org.apache.beam.sdk.schemas.SchemaRegistry;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.TupleTag;
+import org.apache.beam.sdk.values.TypeDescriptor;
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+
+class SqlTransformUtils {
+
+  private SqlTransformUtils() {}
+
+  @SuppressWarnings("unchecked")
+  public static <T> void setCoderForOutput(
+      SchemaRegistry schemaRegistry,
+      CoderRegistry coderRegistry,
+      TupleTag<?> tag,
+      PCollection<?> output,
+      TypeInformation<T> typeInfo) {
+    TypeDescriptor<T> outputTypeDescriptor = (TypeDescriptor<T>) tag.getTypeDescriptor();
+    PCollection<T> out = (PCollection<T>) output;
+    try {
+      out.setSchema(
+          schemaRegistry.getSchema(outputTypeDescriptor),
+          outputTypeDescriptor,
+          schemaRegistry.getToRowFunction(outputTypeDescriptor),
+          schemaRegistry.getFromRowFunction(outputTypeDescriptor));
+    } catch (NoSuchSchemaException e) {
+      try {
+        out.setCoder(coderRegistry.getCoder(outputTypeDescriptor));
```

Comment:
Can we check the type is Beam row or not to set up either the schema coder or the specific type coder? Seems if the user uses the specific types, the schema coder will do extra serde to get to the types.

Reply:
Here is the thing I thought about: after applying the FlinkSQLTransform. there can be two apis to use:

1. use typed api:
 input.apply(FlinkSQLTransform).apply(MapElements.of(....)).apply(Count.perKey())

2. use schema api:
input.apply(FlinkSQLTransform).apply(Select.fieldNames(..)).apply(Group.byField())

This patch will set the coder to use schema coder, so the second case should be supported naturally. I was thinking about the first case. It still should work as normal given Beam can use schema to serde the record for grouping.